policy-controller/k8s/index/src/outbound_index.rs

```diff
@@ -0,0 +1,281 @@
+    policy::{httproute::HttpRouteRule, HttpRoute},
+    ResourceExt, Service,
+};
+use parking_lot::RwLock;
+use std::{fmt, net::IpAddr, num::NonZeroU16, sync::Arc};
+use tokio::sync::watch;
+
+use super::http_route::convert;
+
+pub type SharedIndex = Arc<RwLock<Index>>;
+
+#[derive(Debug)]
+pub struct Index {
+    namespaces: NamespaceIndex,
+    services: HashMap<IpAddr, ServiceRef>,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct ServiceRef {
+    pub name: String,
+    pub namespace: String,
+}
+
+#[derive(Debug)]
+pub struct NamespaceIndex {
+    by_ns: HashMap<String, Namespace>,
+    cluster_domain: Arc<String>,
+}
+
+#[derive(Debug, Default)]
+struct Namespace {
+    services: HashMap<ServicePort, ServiceRoutes>,
+    namespace: Arc<String>,
+    cluster_domain: Arc<String>,
+}
+
+#[derive(Debug, PartialEq, Eq, Hash)]
+struct ServicePort {
+    service: String,
+    port: NonZeroU16,
+}
+
+#[derive(Debug)]
+struct ServiceRoutes {
+    routes: HashMap<String, OutboundHttpRoute>,
+    watch: watch::Sender<OutboundPolicy>,
+    namespace: Arc<String>,
+    cluster_domain: Arc<String>,
+}
+
+impl kubert::index::IndexNamespacedResource<HttpRoute> for Index {
+    fn apply(&mut self, route: HttpRoute) {
+        tracing::debug!(name = route.name_unchecked(), "indexing route");
+        let ns = route.namespace().expect("HttpRoute must have a namespace");
+        self.namespaces
+            .by_ns
+            .entry(ns.clone())
+            .or_insert_with(|| Namespace {
+                services: Default::default(),
+                namespace: Arc::new(ns),
+                cluster_domain: self.namespaces.cluster_domain.clone(),
+            })
+            .apply(route);
+    }
+
+    fn delete(&mut self, namespace: String, name: String) {
+        if let Some(ns_index) = self.namespaces.by_ns.get_mut(&namespace) {
+            ns_index.delete(name);
+        }
+    }
+}
+
+impl kubert::index::IndexNamespacedResource<Service> for Index {
+    fn apply(&mut self, service: Service) {
+        let name = service.name_unchecked();
+        let ns = service.namespace().expect("Service must have a namespace");
+        if let Some(cluster_ip) = service
+            .spec
+            .and_then(|spec| spec.cluster_ip)
+            .filter(|ip| !ip.is_empty() && ip != "None")
+        {
+            match cluster_ip.parse() {
+                Ok(addr) => {
+                    let service_ref = ServiceRef {
+                        name,
+                        namespace: ns,
+                    };
+                    self.services.insert(addr, service_ref);
+                }
+                Err(error) => {
+                    tracing::error!(%error, service=name, "invalid cluster ip");
+                }
+            }
+        }
+    }
+
+    fn delete(&mut self, namespace: String, name: String) {
+        let service_ref = ServiceRef { name, namespace };
+        self.services.retain(|_, v| *v != service_ref);
```

Comment:
this won't delete the `ServiceRoute`s for a service, and therefore won't end any currently outstanding watches for that service. is that intentional?

i would expect that if a Service is deleted, we would want to tell any proxy that's currently watching an IP that corresponds to that Service that the Service no longer exists, so it will revert to the default policy for that IP until we see a new Service assigned that address...